### ISSUE TITLE
ssl CheckCommand does not support SNI

### ISSUE THREAD

**Issue by brandoss** [Icinga/icinga2#5577]
In file, itl/command-plugins.conf

ssl CheckCommand dropped -N argument which is required for SNI to work in commit: https://github.com/Icinga/icinga2/commit/a6978f7b63ff1113530d320cb19b84c46f80cbfa#diff-d35f2508c7085c3303767d379b84e8ff

ssl_address does not work as documented and does not cause check_tcp to use SNI.  This causes the ssl CheckCommand to not always check the correct certificate from a server providing multiple via SNI.

-N is required with check_tcp
```
# /usr/lib64/nagios/plugins/check_tcp --ssl -p 443 -t 10 -D 90,30 -H icinga.org
OK - Certificate '*.icinga.com' will expire on 2019-12-26 18:59 -0500/EST.
TCP OK - 0.421 second response time on icinga.org port 443|time=0.420649s;;;0.000000;10.000000
Returns the icinga.com cert instead of the icinga.org cert

With the -N we get the correct cert.
# /usr/lib64/nagios/plugins/check_tcp --ssl -p 443 -t 10 -D 90,30 -H icinga.org -N icinga.org
OK - Certificate '*.icinga.org' will expire on 2018-07-26 19:59 -0400/EDT.
TCP OK - 0.271 second response time on icinga.org port 443|time=0.270620s;;;0.000000;10.000000

**Comment by mcktr** [comment]
To me it seems that the option *-N* is dropped inside the *check_tcp* plugin. Which version of the plugin are you using? 

```
michael@ganymed:~/Coding/icinga2$ /usr/lib/nagios/plugins/check_tcp --ssl -p 443 -t 10 -D 90,30 -H icinga.org
OK - Certificate '*.icinga.com' will expire on Do 26 Dez 2019 23:59:59 GMT +0000.
TCP OK - 0,111 second response time on icinga.org port 443|time=0,110581s;;;0,000000;10,000000

michael@ganymed:~/Coding/icinga2$ /usr/lib/nagios/plugins/check_tcp --ssl -p 443 -t 10 -D 90,30 -H icinga.org -N icinga.org
/usr/lib/nagios/plugins/check_tcp: Ungültige Option -- N
Usage:
check_tcp -H host -p port [-w <warning time>] [-c <critical time>] [-s <send string>]
[-e <expect string>] [-q <quit string>][-m <maximum bytes>] [-d <delay>]
[-t <timeout seconds>] [-r <refuse state>] [-M <mismatch state>] [-v] [-4|-6] [-j]
[-D <warn days cert expire>[,<crit days cert expire>]] [-S <use SSL>] [-E]

michael@ganymed:~/Coding/icinga2$ /usr/lib/nagios/plugins/check_tcp --version
check_tcp v2.2 (monitoring-plugins 2.2)
```

For monitoring a certificate I think the *check_http* plugin is more suitable. 

```
michael@ganymed:~/Coding/icinga2$ /usr/lib/nagios/plugins/check_http -H icinga.org -C 5,10
OK - Certificate '*.icinga.com' will expire on Do 26 Dez 2019 23:59:59 GMT +0000.
```

**Comment by brandoss** [comment]
/usr/lib64/nagios/plugins/check_tcp --version
check_tcp v2.2.1 (nagios-plugins 2.2.1)

and

 /usr/lib64/nagios/plugins/check_tcp -V
check_tcp v2.1.4 (nagios-plugins 2.1.4)

Please note the output from your check_http is returning the icinga.com cert not icinga.org cert.

**Comment by brandoss** [comment]
Digging into this a little more and the nagios-plugins which is the plugins package for REHL/CENTOS supports SNI with check_tcp and Debian based systems that use monitoring-plugins does not support SNI in check_tcp.

So, I can see how using it could cause Debian systems to break if the config specified ssl_sni.  However, the documentation @ https://www.icinga.com/docs/icinga2/latest/doc/10-icinga-template-library/#ssl does specify the use of ssl_sni.  Which currently does nothing and does not enable SNI for either nagios-plugins or monitoring-plugins.

**Comment by mcktr** [comment]
I searched for differences between nagios-plugins and monitoring-plugins *check_tcp* and found out that the SNI feature is not implemented in the monitoring-plugins. 

I have bit struggle to understand why icinga.com is not the right certificate. All my tests that I did when using icinga.org are immediately forwarded/redirected to icinga.com. So from my point of view the right certificate is icinga.com, because you are getting an 301 redirect to icinga.com.    

Following test makes the thing more confusing 
```
/usr/lib64/nagios/plugins/check_tcp --ssl -p 443 -t 10 -D 90,30 -H icinga.de -N icinga.de
SSL OK - Certificate '*.icinga.com' will expire on 2019-12-26 23:59 +0000/UTC. TCP OK - 0,119 second response time on icinga.de port 443|time=0,119147s;;;0,000000;10,000000
```
How could this be possible, if the right certificate for icinga.org must be icinga.org? 

I'll guess it's hard to getting an correct answer here unless we know the webserver configuration. As far as I can get this together I believe that the webserver redirects all visits to icinga.com, but from icinga.org which was earlier the main domain has still an valid certificate.

But you are right with `ssl_sni` in the documentation, this could be dropped unless while is has no effect on this check command.

**Comment by brandoss** [comment]
icinga.com is not the correct cert because check_tcp should not be following http redirects.  Even with http the first ssl connection is to icinga.org and thus needs a valid cert for icinga.org to send the redirect.  icinga.org does appear to be configured correctly to work with the right cert and SNI.  This can be tested on thousands of domains accross the Internet.

The https for icinga.de appears to be misconfigured on the server.  Go to https://icinga.de and not http://icinga.de and you will be presented with an invalid cert because it is for *.icinga.com

I would prefer to submit a pull request to add -N back, but I understand if it makes more sense for the project to drop ssl_sni from the documentation.

**Comment by mcktr** [comment]
Wow, thanks for the clarification, makes total sense! :+1: 

I send a PR to remove the `ssl_sni` attribute. On Debian based distributions the monitoring-plugins packages came the default (even when you install nagios-plugins), so I guess it is the best to drop this to hold the compatibility for both packages.

What we need is that monitoring-plugins implements SNI to the *check_tcp* plugin. Could you please open an issue over there?

**Comment by dnsmichi** [comment]
Fixed in #5578.
